janusgraph-core/src/main/java/org/janusgraph/graphdb/util/SubqueryIterator.java

```diff
@@ -73,24 +75,37 @@ public SubqueryIterator(JointIndexQuery.Subquery subQuery, IndexSerializer index
 
     @Override
     public boolean hasNext() {
-        if (!elementIterator.hasNext() && currentIds != null) {
-            indexCache.put(subQuery, currentIds);
-            profiler.stopTimer();
-            isTimerRunning = false;
-            profiler.setResultSize(currentIds.size());
+        final boolean hasNext = elementIterator.hasNext();
+        if (!hasNext) {
+            if (currentIds != null) {
+                indexCache.put(subQuery, currentIds);
+            }
+            close();
         }
-        return elementIterator.hasNext();
+        return hasNext;
```

Comment:
Don't forget to cache a query when the result is empty and nobody calls `hasNext` or `next` or `close`.

Reply:
I am a bit against doing so because generally, I want to avoid running the real (and maybe costly) computation in the constructor itself, especially when nobody calls `hasNext` or `next` afterward. In this class, the iterator comes from an unfiltered iterator which intersects with `otherResults`. The intersection can be expensive. Thus, if nobody uses this iterator afterward, the "pre-computation" will be wasted.

That being said, in this class, we know the underlying stream is not lazy, which means in the constructor, the underlying index query already happens. Probably we want to cache all results in the constructor even if nobody uses iterator, or the iterator is used but not exhausted... However, if we don't consider the internal implementation of `stream = indexSerializer.query(subQuery, tx).peek(r -> currentIds.add(r));`, we should consider it as lazy and avoid eager fetching if no one calls `hasNext` or `next` IMO.

Reply:
Great points! Totally makes sense.